clang/tools/clang-format/clang-format.el

```diff
@@ -146,18 +146,115 @@ is a zero-based file offset, assuming ‘utf-8-unix’ coding."
     (lambda (byte &optional _quality _coding-system)
       (byte-to-position (1+ byte)))))
 
-;;;###autoload
-(defun clang-format-region (start end &optional style assume-file-name)
-  "Use clang-format to format the code between START and END according to STYLE.
-If called interactively uses the region or the current statement if there is no
-no active region. If no STYLE is given uses `clang-format-style'. Use
-ASSUME-FILE-NAME to locate a style config file, if no ASSUME-FILE-NAME is given
-uses the function `buffer-file-name'."
-  (interactive
-   (if (use-region-p)
-       (list (region-beginning) (region-end))
-     (list (point) (point))))
-
+(defun clang-format--vc-diff-match-diff-line (line)
+  ;; Matching something like:
+  ;; "@@ -80 +80 @@" or "@@ -80,2 +80,2 @@"
+  ;; Return as "<LineStart>:<LineEnd>"
+  (when (string-match "^@@\s-[0-9,]+\s\\+\\([0-9]+\\)\\(,\\([0-9]+\\)\\)?\s@@$" line)
+    ;; If we have multi-line diff
+    (if (match-string 3 line)
+        (concat (match-string 1 line)
+                ":"
+                (number-to-string
+                 (+ (string-to-number (match-string 1 line))
+                    (string-to-number (match-string 3 line)))))
+      (concat (match-string 1 line) ":" (match-string 1 line)))))
+
+(defun clang-format--vc-diff-get-diff-lines (file-orig file-new)
+  "Return all line regions that contain diffs between FILE-ORIG and
+FILE-NEW.  If there is no diff 'nil' is returned. Otherwise the
+return is a 'list' of lines in the format '--lines=<start>:<end>'
+which can be passed directly to 'clang-format'"
+  ;; Temporary buffer for output of diff.
+  (with-temp-buffer
+    (let ((status (call-process
```

Comment:
Regarding temporary file use: did you check on the `diff-buffers` function? it should be possible to diff two temporary buffers without creating temporary files.

Looking into the implementation `diff-no-select` may be what your after as that looks to be intended for non interactive use.

Reply:
I don't think so, or at least `diff.el` uses temporary files:

```
(defun diff-file-local-copy (file-or-buf)
  "Like `file-local-copy' but also supports a buffer as the argument.
When FILE-OR-BUF is a buffer, return the filename of a local
temporary file with the buffer's contents."
  (if (bufferp file-or-buf)
      (with-current-buffer file-or-buf
        (let ((tempfile (make-temp-file "buffer-content-")))
          (if diff-entire-buffers
              (write-region nil nil tempfile nil 'nomessage)
            (write-region (point-min) (point-max) tempfile nil 'nomessage))
          tempfile))
    (file-local-copy file-or-buf)))
...
  (let* ((old-alt (diff-file-local-copy old))
	 (new-alt (diff-file-local-copy new))
	 (command
	  (mapconcat #'identity
		     `(,diff-command
		       ;; Use explicitly specified switches
		       ,@switches
                       ,@(mapcar #'shell-quote-argument
                                 (nconc
                                  (and (or old-alt new-alt)

```

Reply:
Ah, I wasn't aware of that, even so, there is some advantage in using emacs diff functionality as it handles error cases, removing temp files etc.

Reply:
Okay, I'm opposed to changing to use `diff-no-select`. It is notably slower. I think because we go through temporary buffers. I'm going to push the code that uses `diff-no-select`, unless you see a way to speed it up I think we should keep the current implementation.

Benchmark code:
```
(require 'benchmark)
(benchmark-elapse
  (with-current-buffer "ValueTracking.cpp" ;; 10000 line file
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)
    (clang-format-vc-diff)            
    )
  )
```
Bespoke Diff   : 0.1 += 0.05 sec
`diff-no-select`: 0.2 += 0.05 sec

If I scale `ValueTracking.cpp` to 100000 lines:

Bespoke Diff   : 0.45 += 0.1 sec
`diff-no-select`: 1.1 += 0.1 sec

So it seems twice as slow, I imagine because we end up sending the VC version of the file through memory.

Reply:
NB: These times are with no lines going to clang-format, the call to clang-format itself is dominant (depends on actual code, but on ValueTracking ~5x the cost of the diff setup).

Reply:
Also note even if we stick with temporary file, `diff-no-select` still re-copies the file.

Reply:
In this case, include a code-comment for why `diff-no-select` is not used.